### PR TITLE
Update harness for aws_cryptosdk_enc_ctx_clone

### PR DESCRIPTION
**Pull request by tegansb**
*Description of changes:*
Adds additional assumptions to proof for aws_cryptosdk_enc_ctx_clone_harness. 
Needs new stubs from https://github.com/awslabs/aws-c-common/pull/714. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c [hunk 1]
```diff
@@ -20,19 +20,24 @@
...
-void make_hash_table_with_no_backing_store(struct aws_hash_table *map, size_t max_table_entries);
-
-/**
- * The actual proof
- */
 void aws_cryptosdk_enc_ctx_clone_harness() {
+    /* Nondet Input */
     struct aws_hash_table dest;
-    make_hash_table_with_no_backing_store(&dest, MAX_NUM_ELEMS);
+    struct aws_hash_table src;
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c:26]: `struct aws_hash_table* src = malloc(sizeof(*src))`

**Review by danielsn — CHANGES_REQUESTED** [review]

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c [hunk 2]
```diff
@@ -20,19 +20,24 @@
...
+
+    /* Assumptions */
+    ensure_allocated_hash_table(&dest, MAX_TABLE_SIZE);
+    __CPROVER_assume(dest.p_impl != NULL);
+    __CPROVER_assume(dest.p_impl->entry_count <= MAX_TABLE_SIZE);
```
[the remark this answers is not in the document]
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c:31]: Nope, I couldn't find anything ensuring this condition which is needed for the loop unwinding. I also haven't come across another proof needing this condition which is why I didn't make an auxiliary function for it.

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c (hunk 1 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c]: Updated!

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/Makefile [hunk 3]
```diff
@@ -20,7 +20,7 @@ include ../Makefile.local_default
 
 #########
 # Local vars
-MAX_NUM_ELEMS = 4
+MAX_TABLE_SIZE ?= 2
 # Time on my laptop: 4m50s
```
[the remark this answers is not in the document]
Comment (danielsn) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/Makefile:24]: Is the time comment still correct?

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c [hunk 4]
```diff
@@ -20,19 +20,24 @@
...
-    __CPROVER_assume(aws_hash_table_is_valid(&src));
+    ensure_allocated_hash_table(src, MAX_TABLE_SIZE);
+    __CPROVER_assume(aws_hash_table_is_valid(src));
+    __CPROVER_assume(src->p_impl != NULL);
+    __CPROVER_assume(src->p_impl->entry_count <= MAX_TABLE_SIZE);
+    ensure_hash_table_has_valid_destroy_functions(src);
 
-    int rval = aws_cryptosdk_enc_ctx_clone(can_fail_allocator(), &dest, &src);
+    /* Operation under verification */
+    int rval = aws_cryptosdk_enc_ctx_clone(can_fail_allocator(), dest, src);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c:42]: Not blocking for this PR, but add an issue to add postcondition checks here.

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c [hunk 5]
```diff
@@ -20,19 +20,24 @@
...
+    __CPROVER_assume(dest->p_impl != NULL);
+    __CPROVER_assume(dest->p_impl->entry_count <= MAX_TABLE_SIZE);
+    ensure_hash_table_has_valid_destroy_functions(dest);
 
-    struct aws_hash_table src;
-    make_hash_table_with_no_backing_store(&src, MAX_NUM_ELEMS);
-    __CPROVER_assume(aws_hash_table_is_valid(&src));
+    ensure_allocated_hash_table(src, MAX_TABLE_SIZE);
+    __CPROVER_assume(aws_hash_table_is_valid(src));
+    __CPROVER_assume(src->p_impl != NULL);
```
Comment (danielsn) [inline .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c:37]: Is this needed? Can it be valid with a null impl?

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c (hunk 4 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c:42]: Issue added! https://github.com/aws/aws-encryption-sdk-c/issues/613

File: .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c (hunk 5 above)
Comment (tegansb) [reply .cbmc-batch/jobs/aws_cryptosdk_enc_ctx_clone/aws_cryptosdk_enc_ctx_clone_harness.c]: Nope, not needed! It's removed now

**Review by danielsn — APPROVED** [review]

**Review by alex-chew — APPROVED** [review]

**Pull request merged**